src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -30,16 +30,9 @@ protected Operation(SemanticModel? semanticModel, SyntaxNode syntax, bool isImpl
 #if DEBUG
             if (semanticModel != null)
             {
-                Debug.Assert(semanticModel.ContainingModelOrSelf != null);
-                if (semanticModel.Language != LanguageNames.CSharp && semanticModel.IsSpeculativeSemanticModel)
-                {
-                    Debug.Assert(semanticModel.ContainingModelOrSelf == semanticModel);
-                }
-                else
-                {
-                    Debug.Assert(semanticModel.ContainingModelOrSelf != semanticModel);
-                    Debug.Assert(semanticModel.ContainingModelOrSelf.ContainingModelOrSelf == semanticModel.ContainingModelOrSelf);
-                }
+                Debug.Assert(semanticModel.ContainingPublicModelOrSelf != null);
+                Debug.Assert(semanticModel.ContainingPublicModelOrSelf != semanticModel);
```

Comment:
nit: Is it actually important that this not be a public semantic model?

Reply:
> Is it actually important that this not be a public semantic model?

I think implementation actually depends on that somewhere. I don't remember exact details, but, for example, operation nodes are cached by the internal semantic model. In any case, this was a legacy behavior and I do not intend to change it in this PR.